diff -Nur ipsec-tools-0.6.beta1/src/racoon/racoonctl.c ipsec-tools-0.6.beta1.org/src/racoon/racoonctl.c
--- ipsec-tools-0.6.beta1/src/racoon/racoonctl.c	2004-12-30 12:08:32.000000000 +0100
+++ ipsec-tools-0.6.beta1.org/src/racoon/racoonctl.c	2005-06-12 15:44:40.201749016 +0200
@@ -80,8 +80,6 @@
 #include "ipsec_doi.h"
 #include "evt.h"
 
-static char *adminsock_path = ADMINSOCK_PATH;
-
 static void usage __P((void));
 static vchar_t *get_combuf __P((int, char **));
 static int handle_recv __P((vchar_t *));
@@ -260,7 +258,7 @@
 	 */
 	setenv("POSIXLY_CORRECT", "1", 0);
 #endif
-	while ((c = getopt(ac, av, "lds:")) != -1) {
+	while ((c = getopt(ac, av, "ld:")) != -1) {
 		switch(c) {
 		case 'l':
 			long_format++;
@@ -270,10 +268,6 @@
 			loglevel++;
 			break;
 
-		case 's':
-			adminsock_path = optarg;
-			break;
-
 		default:
 			usage();
 			exit(0);
