--- ipsec-tools-0.2.2.orig/src/libipsec/pfkey.c	2003-03-03 15:56:54.000000000 -0800
+++ ipsec-tools-0.2.2.pfkey/src/libipsec/pfkey.c	2003-11-03 19:54:07.770143564 -0800
@@ -1181,7 +1181,7 @@
 		+ sizeof(struct sadb_lifetime)
 		+ sizeof(struct sadb_lifetime);
 
-	if (e_type != SADB_EALG_NONE)
+	if (e_type != SADB_EALG_NONE && satype != SADB_X_SATYPE_IPCOMP)
 		len += (sizeof(struct sadb_key) + PFKEY_ALIGN8(e_keylen));
 	if (a_type != SADB_AALG_NONE)
 		len += (sizeof(struct sadb_key) + PFKEY_ALIGN8(a_keylen));
@@ -1221,7 +1221,7 @@
 		return -1;
 	}
 
-	if (e_type != SADB_EALG_NONE) {
+	if (e_type != SADB_EALG_NONE && satype != SADB_X_SATYPE_IPCOMP) {
 		p = pfkey_setsadbkey(p, ep, SADB_EXT_KEY_ENCRYPT,
 		                   keymat, e_keylen);
 		if (!p) {
