diff -urN ipsec-tools-0.2.2.org/src/racoon/configure.in ipsec-tools-0.2.2/src/racoon/configure.in
--- ipsec-tools-0.2.2.org/src/racoon/configure.in	2003-07-12 01:20:51.631425000 +0200
+++ ipsec-tools-0.2.2/src/racoon/configure.in	2003-07-12 01:22:45.830064168 +0200
@@ -136,7 +136,7 @@
 
 AC_MSG_CHECKING(if you have KAME tcpdump in $srcdir/../tcpdump)
 if test -f "$srcdir/../tcpdump/print-isakmp.c"; then
-	LIBOBJS="$LIBOBJS print-isakmp.o"
+	AC_LIBOBJ([print-isakmp])
 	AC_DEFINE(HAVE_PRINT_ISAKMP_C)
 	AC_MSG_RESULT(yes)
 else
@@ -144,7 +144,7 @@
 fi
 
 AC_MSG_CHECKING(if __func__ is available)
-AC_TRY_COMPILE(dnl
+AC_TRY_COMPILE(
 [#include <stdio.h>
 ], [char *x = __func__;],
 	[AC_DEFINE(HAVE_FUNC_MACRO)
@@ -199,7 +199,7 @@
 		case $i in
 		inria)
 			dnl http://www.kame.net/
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef IPV6_INRIA_VERSION
 yes
@@ -212,7 +212,7 @@
 			;;
 		kame)
 			dnl http://www.kame.net/
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef __KAME__
 yes
@@ -240,7 +240,7 @@
 			fi
 			;;
 		toshiba)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <sys/param.h>
 #ifdef _TOSHIBA_INET6
 yes
@@ -251,7 +251,7 @@
 				CPPFLAGS="-DINET6 $CPPFLAGS"])
 			;;
 		v6d)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include </usr/local/v6/include/sys/v6config.h>
 #ifdef __V6D__
 yes
@@ -262,7 +262,7 @@
 				CPPFLAGS="-I/usr/local/v6/include $CPPFLAGS"])
 			;;
 		zeta)
-			AC_EGREP_CPP(yes, [dnl
+			AC_EGREP_CPP(yes, [
 #include <sys/param.h>
 #ifdef _ZETA_MINAMI_INET6
 yes
@@ -298,7 +298,7 @@
 
 if test "$ipv6" = "yes"; then
 	AC_MSG_CHECKING(for advanced API support)
-	AC_CACHE_VAL(racoon_cv_advapi, [dnl
+	AC_CACHE_VAL(racoon_cv_advapi, [
 	AC_TRY_COMPILE([#ifndef INET6
 #define INET6
 #endif
@@ -528,7 +528,7 @@
 
 dnl PF_KEYv2 support check
 AC_MSG_CHECKING(for PF_KEYv2 support)
-AC_CACHE_VAL(ac_cv_pfkey, [dnl
+AC_CACHE_VAL(ac_cv_pfkey, [
 AC_TRY_COMPILE([
 #include <sys/types.h>
 #include <sys/socket.h>
@@ -579,7 +579,7 @@
 	case $i in
 	inria)
 		dnl http://www.kame.net/
-		AC_EGREP_CPP(yes, [dnl
+		AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef IPV6_INRIA_VERSION
 yes
@@ -590,7 +590,7 @@
 		;;
 	kame)
 		dnl http://www.kame.net/
-		AC_EGREP_CPP(yes, [dnl
+		AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef __KAME__
 yes
@@ -600,7 +600,7 @@
 			ipseclibdir=/usr/local/v6/lib])
 		;;
 	linux)
-		AC_EGREP_CPP(yes, [dnl
+		AC_EGREP_CPP(yes, [
 #include <netinet/in.h>
 #ifdef __linux__
 yes
