diff -Nur ipsec-tools-0.6.4-orig/src/racoon/cfparse.y ipsec-tools-0.6.4/src/racoon/cfparse.y
--- ipsec-tools-0.6.4-orig/src/racoon/cfparse.y	2005-10-17 18:23:50.000000000 +0200
+++ ipsec-tools-0.6.4/src/racoon/cfparse.y	2005-12-28 14:07:51.000000000 +0100
@@ -821,9 +821,7 @@
 					return -1;
 				}
 				$$ = ipsecdoi_sockaddr2id(saddr,
-					$3 == (sizeof(struct in_addr) << 3) &&
-						$1 == IDTYPE_ADDRESS
-					  ? ~0 : $3,
+					$3 == ~0 ? (sizeof(struct in_addr) << 3): $3,
 					$5);
 				break;
 #ifdef INET6
@@ -834,9 +832,7 @@
 					return -1;
 				}
 				$$ = ipsecdoi_sockaddr2id(saddr,
-					$3 == (sizeof(struct in6_addr) << 3) &&
-						$1 == IDTYPE_ADDRESS
-					  ? ~0 : $3,
+					$3 == ~0 ? (sizeof(struct in6_addr) << 3) : $3,
 					$5);
 				break;
 #endif
diff -Nur ipsec-tools-0.6.4-orig/src/racoon/cftoken.l ipsec-tools-0.6.4/src/racoon/cftoken.l
--- ipsec-tools-0.6.4-orig/src/racoon/cftoken.l	2005-11-06 18:18:26.000000000 +0100
+++ ipsec-tools-0.6.4/src/racoon/cftoken.l	2005-12-28 14:07:51.000000000 +0100
@@ -471,7 +471,6 @@
 fqdn		{ YYD; yylval.num = IDTYPE_FQDN; return(IDENTIFIERTYPE); }
 keyid		{ YYD; yylval.num = IDTYPE_KEYID; return(IDENTIFIERTYPE); }
 address		{ YYD; yylval.num = IDTYPE_ADDRESS; return(IDENTIFIERTYPE); }
-subnet		{ YYD; yylval.num = IDTYPE_SUBNET; return(IDENTIFIERTYPE); }
 asn1dn		{ YYD; yylval.num = IDTYPE_ASN1DN; return(IDENTIFIERTYPE); }
 certname	{ YYD; yywarn("certname will be obsoleted in near future."); yylval.num = IDTYPE_ASN1DN; return(IDENTIFIERTYPE); }
 
diff -Nur ipsec-tools-0.6.4-orig/src/racoon/ipsec_doi.c ipsec-tools-0.6.4/src/racoon/ipsec_doi.c
--- ipsec-tools-0.6.4-orig/src/racoon/ipsec_doi.c	2005-10-17 18:23:50.000000000 +0200
+++ ipsec-tools-0.6.4/src/racoon/ipsec_doi.c	2005-12-28 14:07:51.000000000 +0100
@@ -3742,7 +3742,7 @@
 	switch (saddr->sa_family) {
 	case AF_INET:
 		len1 = sizeof(struct in_addr);
-		if (prefixlen == ~0) {
+		if (prefixlen == (sizeof(struct in_addr) << 3)) {
 			type = IPSECDOI_ID_IPV4_ADDR;
 			len2 = 0;
 		} else {
@@ -3755,7 +3755,7 @@
 #ifdef INET6
 	case AF_INET6:
 		len1 = sizeof(struct in6_addr);
-		if (prefixlen == ~0) {
+		if (prefixlen == (sizeof(struct in6_addr) << 3)) {
 			type = IPSECDOI_ID_IPV6_ADDR;
 			len2 = 0;
 		} else {
diff -Nur ipsec-tools-0.6.4-orig/src/racoon/ipsec_doi.h ipsec-tools-0.6.4/src/racoon/ipsec_doi.h
--- ipsec-tools-0.6.4-orig/src/racoon/ipsec_doi.h	2005-10-17 18:23:50.000000000 +0200
+++ ipsec-tools-0.6.4/src/racoon/ipsec_doi.h	2005-12-28 14:07:51.000000000 +0100
@@ -180,7 +180,6 @@
 #define IDTYPE_ADDRESS		4
 #define IDTYPE_ASN1DN		5
 #define IDTYPE_LOGIN		6
-#define IDTYPE_SUBNET		7
 
 /* The use for checking proposal payload. This is not exchange type. */
 #define IPSECDOI_TYPE_PH1	0
