--- ipsec.org/src/racoon/Makefile.in.org	2004-11-06 16:51:08.511698693 +0100
+++ ipsec/src/racoon/Makefile.in	2004-11-06 16:51:21.263966818 +0100
@@ -127,8 +127,8 @@
 
 install:
 	-mkdir -p $(DESTDIR)$(sbindir)
-	$(INSTALL) -s $(INSTALL_OPTS) -m 555 racoon $(DESTDIR)$(sbindir)
-	$(INSTALL) -s $(INSTALL_OPTS) -m 555 plainrsa-gen $(DESTDIR)$(sbindir)
+	$(INSTALL) $(INSTALL_OPTS) -m 555 racoon $(DESTDIR)$(sbindir)
+	$(INSTALL) $(INSTALL_OPTS) -m 555 plainrsa-gen $(DESTDIR)$(sbindir)
 	-mkdir -p $(DESTDIR)$(mandir)/man8/
 	$(INSTALL) $(INSTALL_OPTS) -m 444 $(srcdir)/racoon.8 $(DESTDIR)$(mandir)/man8
 	$(INSTALL) $(INSTALL_OPTS) -m 444 $(srcdir)/plainrsa-gen.8 $(DESTDIR)$(mandir)/man8
