--- ipsec-tools-0.8.2/src/setkey/setkey.c~	2012-08-23 13:10:46.000000000 +0200
+++ ipsec-tools-0.8.2/src/setkey/setkey.c	2015-08-09 20:01:16.128401541 +0200
@@ -40,7 +40,6 @@
 #include <sys/socket.h>
 #include <sys/time.h>
 #include <sys/stat.h>
-#include <sys/sysctl.h>
 #include <err.h>
 #include <netinet/in.h>
 #include <net/pfkeyv2.h>
--- ipsec-tools-0.8.2/src/racoon/pfkey.c~	2012-08-23 13:10:45.000000000 +0200
+++ ipsec-tools-0.8.2/src/racoon/pfkey.c	2015-08-09 20:01:56.678932946 +0200
@@ -59,7 +59,6 @@
 #include <sys/param.h>
 #include <sys/socket.h>
 #include <sys/queue.h>
-#include <sys/sysctl.h>
 
 #include <net/route.h>
 #include <net/pfkeyv2.h>
